lib/vector-core/src/stream/batcher.rs

```diff
@@ -0,0 +1,538 @@
+        self.stream.size_hint()
+    }
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut this = self.project();
+        loop {
+            if !this.closed_batches.is_empty() {
+                return Poll::Ready(this.closed_batches.pop());
+            }
+            match this.stream.as_mut().poll_next(cx) {
+                Poll::Pending => {
+                    if this.timer.has_elapsed() {
+                        this.closed_batches.extend(
+                            this.batches
+                                .drain()
+                                .map(|(key, batch)| (key, batch.destruct().1)),
+                        );
+                        this.timer.reset();
+                    }
+                    return Poll::Pending;
+                }
+                Poll::Ready(None) => {
+                    // Now that the underlying stream is closed we need to clear
+                    // out our batches.
+                    if !this.batches.is_empty() {
+                        this.closed_batches.extend(
+                            this.batches
+                                .drain()
+                                .map(|(key, batch)| (key, batch.destruct().1)),
+                        );
+                        continue;
+                    }
+                    return Poll::Ready(None);
+                }
+                Poll::Ready(Some(item)) => {
+                    let item: Prt::Item = item.into();
+                    let item_key = this.partitioner.partition(&item);
+
+                    if let Some(batch) = this.batches.get_mut(&item_key) {
+                        if batch.has_space(&item) {
+                            batch.push(item);
+                        } else {
+                            let new_batch =
+                                Batch::new(*this.batch_item_limit, *this.batch_allocation_limit)
+                                    .with(item);
+                            let batch = mem::replace(batch, new_batch);
+                            let (_, arr) = batch.destruct();
+                            if this.timer.has_elapsed() {
+                                this.closed_batches.extend(
+                                    this.batches
+                                        .drain()
+                                        .map(|(key, batch)| (key, batch.destruct().1)),
+                                );
+                                this.timer.reset();
+                            }
+                            return Poll::Ready(Some((item_key, arr)));
+                        }
+                    } else {
+                        let batch =
+                            Batch::new(*this.batch_item_limit, *this.batch_allocation_limit)
+                                .with(item);
+                        this.batches.insert(item_key, batch);
+                    }
```

Comment:
This could be a little more succinct if you use the entry API, since we're getting an owned partition key no matter what in most cases.

Reply:
Good call.

Reply:
Hmm so I fiddled with this for a minute and didn't come out ahead. Ends up looking like `this.batches.entry(item_key).or_insert(Batch::new(...))` but isn't any shorter.

Reply:
Yeah, now that I just tried to write it out myself, it's not really possibly to same much given that you're trying to exchange the batch at the same time if it ends up being full.  Drats.